### ISSUE TITLE
Assertion Failed!: Cannot call EnqueueAction() with a non Action parameter. : false

### ISSUE THREAD

**Issue by angelsalazar** [forcedotcom/aura#135]
Hello, 
I hope you can help me to figure out why I am getting the following error:

`Assertion Failed!: Cannot call EnqueueAction() with a non Action parameter. : false`

My code structure is the following
I have a Component from which other components extends in order to inherit the attributes and helper methods

LngService.cmp
```
<!-- PARENT SERVICE COMPONENT -->
<aura:component abstract="true" 
    extensible="true">
	<!-- attributes -->
	<!-- redering variables -->
	<aura:attribute name="creating" type="Boolean" default="false" access="public"/>
	<aura:attribute name="updating" type="Boolean" default="false" access="public"/>
	<aura:attribute name="deleting" type="Boolean" default="false" access="public"/>
	<aura:attribute name="loading" type="Boolean" default="false" access="public"/>
	<aura:attribute name="ready" type="Boolean" default="false" access="public"/>
</aura:component>
```
LngServiceHelper.js
```
({
	executeCallback : function (callback) {
		if (!$A.util.isUndefinedOrNull(callback) && 
			$A.util.isFunction(callback)) {
			callback();
		}
	},
	displayToast : function (opts) {
		var options = opts || {
			title : 'No title provided',
			message : 'No message provided'
		};

		var toastEvent = $A.get("e.force:showToast");
		
		if ($A.util.isUndefinedOrNull(toastEvent)) 
			return console.warn('toast is not supported');

		toastEvent.setParams(options).fire()
	},
	apexRequest : function (component, actionName, params) {
        return new Promise (function (resolve, reject) {
            var action = component.get(['c', actionName].join('.'));
            if (params && typeof params === 'object') {
                action.setParams(params);
            }

            action.setCallback(this, cb);

            function cb (response) {
                console.timeEnd();
                var state = response.getState();
                if (state === 'SUCCESS' && component.isValid()) {
                    resolve(response.getReturnValue());
                } else {
                    reject(response.getError());
                }
            }

            console.time();
            try {
                $A.enqueueAction(action);
            } catch (e) {
                reject(e);
            }
            
        }.bind(this));
    }
})
```
The purpose of this component is to be used as parent for other service components, so I do not have to write the above methods over and over again.
It works fine for other service components that extends from it.
I just can not figure out why it does not work in the following case:

JobApplicantSearch_Service.cmp
```
<aura:component controller="ApplicantTrackingAura"
	extends="c:LngService">
	<!-- attributes -->
	<aura:attribute name="applicants" type="Object[]" access="public" />
	<aura:attribute name="positionIds" type="String[]" access="public" />


	<aura:attribute name="searching" type="Boolean" access="public" />
	<aura:attribute name="searched" type="Boolean" access="public" />
	
	<!-- pagination -->
	<aura:attribute name="page" type="Integer" access="public" />
	<aura:attribute name="pages" type="Integer" access="public" />
	<aura:attribute name="total" type="Integer" access="public" />

	<!-- methods -->
	<aura:method name="searchApplicants" action="{!c.handleSearchApplicants}">
		<aura:attribute name="positionsSearch" type="Object" required="true" />
		<aura:attribute name="callback" type="Function" required="true" />
	</aura:method>
	<!-- handlers -->
</aura:component>
```
JobApplicantSearch_ServiceController.js
```
({
	handleSearchApplicants : function (component, event, helper) {
		var args   = event.getParam('arguments'),
			query  = args.positionsSearch,
			cb     = args.callback;

		
		helper.searchApplicants(component, query, cb);
	}
})
```

JobApplicantSearch_ServiceHelper.js
```
({
	searchApplicants : function (component, query, callback) {
		component.set('v.searching', true);
        
        this.apexRequest(component, 'getJobApplicationsByPositionsv2', {
            searchDataString : JSON.stringify(query)
        })
        .then(this.handleSearchResponse(component, callback))
        .catch(this.handleSearchError(component, callback));
        
		
	},
    handleSearchResponse : function (component, callback) {
        return $A.getCallback(function (response) {
            component.set('v.searching', false);
            response = JSON.parse(response);
            console.log(response);

            if (!response.success) {
                this.displayToast({
                    title : 'Error',
                    message : response.error,
                    type : 'error'
                });
            } else {
                var page     = response.payload.applications.page,
                    pageSize = response.payload.applications.pageSize,
                    total    = response.payload.applications.total,
                    pages    =  Math.ceil(total/pageSize);

                component.set('v.page', page);
                component.set('v.pages',  pages);
                component.set('v.total', total);
                component.set('v.searched', true);
                component.set('v.applicants', response.payload.applications.data);
            }

            this.executeCallback(callback);
        }.bind(this));
    },
    handleSearchError : function (component, callback) {
        return $A.getCallback(function (err) {
            component.set('v.searching', false);
            this.executeCallback(callback);
            console.log(err);
            this.displayToast({
                title : 'Error',
                message : 'Internal Error - Please Contact the Support Team',
                type : 'error'
            });
        }.bind(this));
    }
})
```

ApplicantTrackingAura.cls
```
@auraEnabled
public static String getJobApplicationsByPositionsv2 (String searchDataString) 
	{
		LightningResponse res = new LightningResponse();

		JobApplications.JobApplicationSearch searchData = (
			(JobApplications.JobApplicationSearch)JSON.deserialize(
				searchDataString, 
				JobApplications.JobApplicationSearch.class
			)
		);
		
		if (searchData.positionIds == null ||
			searchData.positionIds.isEmpty()) {
			res.setError('positionIds is required');
			return res.stringify();
		}

		try {
			Paginator.PaginateResponse applications = JobApplicationService.getJobApplicationsByPositionsv2(
				searchData
			);
			res.add('applications', applications);
		} catch (Exception e) {
			System.debug('@@Q - getJobApplicationsByPositionsv2 :' + e.getMessage());
			System.debug('@@Q - getJobApplicationsByPositionsv2 :' + e.getLineNumber());
			System.debug('@@Q - getJobApplicationsByPositionsv2 :' + e.getStackTraceString());

			res.setError(e.getMessage());
		}
		
		return res.stringify();	
	}
```

when debugging

`$A.enqueueAction(action);` throws (I got this after wrapping enqueueAction in a try/catch block, before that I just got `b.X is not a function` )

`Assertion Failed!: Cannot call EnqueueAction() with a non Action parameter. : false `

**Comment by lukeis** [comment]
when you were debugging did you see what the "action" variable is at that point? The error is due to the value passed in to enqueueAction not being an [instanceof Action.](https://github.com/forcedotcom/aura/blob/master/aura-impl/src/main/resources/aura/util/Util.js#L2403-L2405)

**Comment by angelsalazar** [comment]
Hi @lukeis , I just add the following line of code in the method called `apexRequest`

`console.log('%c is Action: ' + $A.util.isAction(action), 'color: blue');`

Resulting in 

`is Action: false`

when using apexRequest method in the child service JobApplicantSearch_Service, but
when using it in other child services, it logs `is Action: true` (the other child services extends form LngService and have the same pattern).

why does the following the line of code not result in a valid action for this specific case?

`var action = component.get(['c', actionName].join('.'));`

I double checked typos and everything I did seems well written.

**Comment by lukeis** [comment]
what is `action` in this case though? Is it undefined / null? or something else?

**Comment by angelsalazar** [comment]
![screen shot 2017-10-17 at 12 01 22 pm](https://user-images.githubusercontent.com/7455593/31678203-00e01d14-b333-11e7-8ad4-fd8e7347c32c.png)

**Comment by angelsalazar** [comment]
Sorry I accidentally close the issue.
this is what I got from `action `

![screen shot 2017-10-17 at 12 01 22 pm](https://user-images.githubusercontent.com/7455593/31678280-4bd97ebe-b333-11e7-84f4-448e25b48d80.png)

**Comment by angelsalazar** [comment]
This is how action looks for other child services

![screen shot 2017-10-17 at 12 06 20 pm](https://user-images.githubusercontent.com/7455593/31678430-cb45c5cc-b333-11e7-88cd-12edd4b71776.png)

**Comment by lukeis** [comment]
do you have locker service enabled? This may be a bug related to that

**Comment by angelsalazar** [comment]
I checked my developer org and I found

> LockerService Auto-Activation
> In Summer '17, LockerService is auto-activated for all orgs with no option to disable.

I am working in Winter '18, I assume LockerService is on.
all my components have api version 40

**Comment by jsbulldog89** [comment]
@angelsalazar So this may or may not be helpful, but I'm also running into this error message with a similar setup. My code is complicated so I've been trying to repro it with a more generic implementation, but can't. But in my case, I figured out the issue so I'm going to pose a few suggestions:

When my components are on 40.0+, the `action` variable outputs what you are seeing for JobApplicantSearch_Service. For components on a lower version, the `action` variable outputs similar to what you are seeing for your other components.
- Verify that all of your components are actually on the same version - either 40.0 for locker service or a lower version to run without locker service.
- In my case, I received the `Assertion Failed!: Cannot call EnqueueAction() with a non Action parameter. : false` error after changing the -meta.xml file of all of my lightning components to 40.0 in my IDE and deploying the changes. There seems to be some sort of caching going on in my browser (Chrome) that even the suggestions listed [on this SE post](https://salesforce.stackexchange.com/questions/134111/salesforce-lightning-component-cache) didn't seem to fix.
- By going in incognito mode, I was able to test the same components and did not receive the error. I went back to normal mode in Chrome, cleared the browser cache, and then did not receive the error either.
